lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2149,6 +2150,76 @@ readNodeTipStateForTxWrite netLayer = do
             $ ErrNodeNotYetInRecentEra nopp
         Right pp -> pure (pp, timeTranslation)
 
+-- | Wallet-specific wrapped version of 'Write.balanceTx', made for the new tx
+-- workflow with Shelley- and Shared- wallet flavors.
+--
+-- Changes to the change state are not written to the DB.
+balanceTx
+    :: forall s era.
+        ( GenChange s
+        , WalletFlavor s
+        , HasSNetworkId (NetworkOf s)
+        , Write.IsRecentEra era
+        , Including AllFlavors '[ 'ShelleyF, 'SharedF] (FlavorOf s)
+        )
+    => WalletLayer IO s
+    -> Write.PParams era
+    -> TimeTranslation
+    -> PartialTx era
+    -> IO (Write.Tx era)
+balanceTx wrk pp timeTranslation partialTx = do
+    (utxo, wallet, _txs) <- liftIO $ readWalletUTxO wrk
+    -- FIXME: Why are we not using the availableUTxO here?
+    let utxoIndex =
+            Write.constructUTxOIndex $
+            Write.fromWalletUTxO utxo
+
+    let utxoAssumptions = case walletFlavor @s of
+            ShelleyWallet -> AllKeyPaymentCredentials
+            SharedWallet -> AllScriptPaymentCredentialsFrom
+                (Shared.paymentTemplate (getState wallet))
+                (sharedWalletScriptLookup (getState wallet)
+                    . Convert.toWalletAddress)
+
```

Comment:
Since we are in `IO`, we will be able to query the node for extra `UTxO` info here later.